src/main/java/io/confluent/idesidecar/restapi/cache/Clients.java

```diff
@@ -0,0 +1,134 @@
+package io.confluent.idesidecar.restapi.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.CaffeineSpec;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import io.confluent.idesidecar.restapi.connections.ConnectionState;
+import io.confluent.idesidecar.restapi.events.Lifecycle;
+import io.quarkus.logging.Log;
+import jakarta.enterprise.event.ObservesAsync;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+/**
+ * Utilities to obtain and cache clients for a given connection and client ID.
+ */
+public abstract class Clients<T extends AutoCloseable> {
+
+  /**
+   * Caffeine spec to use for the client cache. By default, this will be an empty spec.
+   * See the
+   * <a href="https://github.com/ben-manes/caffeine/wiki/Specification">Caffeine Spec</a>
+   * for more information on the format. Inherited classes may override this method
+   * and return a different spec.
+   */
+  private final CaffeineSpec caffeineSpec;
+
+  /**
+   * Store an instance of a Caffeine cache for each connection. The cache will store
+   * clients by client ID and its policy may be configured by the {@link #caffeineSpec}.
+   */
+  private final Map<String, Cache<String, T>> clientsByIdByConnections =
+      new ConcurrentHashMap<>();
+
+  protected Clients(CaffeineSpec caffeineSpec) {
+    this.caffeineSpec = caffeineSpec;
+  }
+
+  protected Clients() {
+    this.caffeineSpec = CaffeineSpec.parse("");
```

Comment:
Maybe it doesn't matter in this simple case, but it's often helpful to have one constructor that is used by others, so that initialization logic is centralized.
```suggestion
    this(CaffeineSpec.parse(""));
```

Reply:
Thanks! Will keep in mind.